File: advent-of-code/2021/day3-1.py
def intify(l):
    n = len(l)
    x = 0

    for i, e in enumerate(l):
        x += e * 2 ** (n - i - 1)

    return x

def solve(xs):
    n = len(xs)
    m = len(xs[0])

    gamma = []
    epsilon = []

    for i in range(m):
        num_ones = 0
        for j in range(n):
            if xs[j][i] == '1':
                num_ones += 1

        gamma.append(0 + (num_ones > n / 2))
        epsilon.append(1 - (num_ones > n / 2))

    return intify(gamma) * intify(epsilon)
            
def main():
    with open("input/3.data") as f:
        xs = list(map(lambda line: line.strip(), f.readlines()))
        print(solve(xs))

if __name__ == "__main__":
    main()

File: advent-of-code/2021/day3-1.hs
import Data.List

strToIntArr :: String -> [Int]
strToIntArr s = [if x == '0' then 0 else 1 | x <- s]

binToDec :: [Int] -> Int
binToDec xs = foldl (\acc x -> 2 * acc + x) 0 xs

mostCommon :: [Int] -> Int
mostCommon xs = if zerosCount > onesCount then 0 else 1
  where
    count x    = length . filter (== x)
    zerosCount = count 0 xs
    onesCount  = count 1 xs

solve :: [[Int]] -> Int
solve xs = gamma * epsilon
  where
    columnBits      = transpose xs
    mostCommonBits  = map mostCommon columnBits
    leastCommonBits = map (\x -> 1 - x) mostCommonBits
    gamma           = binToDec mostCommonBits
    epsilon         = binToDec leastCommonBits

main :: IO ()
main = do
  input <- readFile "input/3.data"
  putStrLn $ (show . solve) (map strToIntArr (lines input))
